juju/juju
# Facade registration cleanup

## Description of change

This change makes a number of improvements to the way API facades are registered. Highlights:

* The global API facade registry has been removed. This avoids all the awkwardness and danger around globals and allows for more flexibility for upcoming projects.
* The handling of feature flags has been removed from the facade registry type. This is no longer necessary and simplifies the registry implementation.
* Fixed error handling in the facade registration infrastructure.

New facades now just require a call be added to `apiserver.AllFacades`. 

A follow-up PR will tidy up the groupings of facades in `AllFacades`.

## QA steps

Unit tests pass. 

Sanity check:

```
$ juju bootstrap dev 
...
$ juju status
Model    Controller  Cloud/Region  Version
default  dev         dev           2.2-beta1.1

App  Version  Status  Scale  Charm  Store  Rev  OS  Notes

Unit  Workload  Agent  Machine  Public address  Ports  Message

Machine  State  DNS  Inst id  Series  AZ  Message

$ juju deploy cs:~cmars/mattermost                        
Located charm "cs:~cmars/mattermost-16".
Deploying charm "cs:~cmars/mattermost-16".
Deployment under prior agreement to terms: isrg-lets-encrypt
menno@maiwa ~/go/src/github.com/juju/juju ± apiserver-facade-cleanup 
$ watch --color juju status --color
# Deploys fine...
$ juju list-resources mattermost
[Service]
Resource  Supplied by  Revision
bdist     charmstore   3

```

## Documentation changes

N.A.

## Bug reference

N.A.
